Detectors/PHOS/simulation/src/Detector.cxx

```diff
@@ -0,0 +1,1029 @@
+//  mHits.clear();   
+  mCurrentTrackID=-1 ;    
+  mCurrentCellID=-1 ; 
+  mCurentSuperParent=-1 ;   
+  mCurrentHit=nullptr ;          
+}
+
+void Detector::Register(){
+  FairRootManager::Instance()->RegisterAny(addNameTo("Hit").data(), mHits, kTRUE);
+}
+Bool_t Detector::ProcessHits(FairVolume* v)
+{
+ //1. Remember all particles first entered PHOS (active medium)
+ //2. Collect all energy depositions in Cell by all secondaries from particle first entered PHOS  
+
+  auto* mcapp = TVirtualMC::GetMC();
+
+  // Check if this is first entered PHOS particle ("SuperParent")
+  TVirtualMCStack * stack = mcapp->GetStack();
+  const Int_t partID=stack->GetCurrentTrackNumber() ;
+  Int_t superParent=-1; 
+  Bool_t isNewPartile = false ;  //Create Hit even if zero energy deposition
+  if(partID != mCurrentTrackID){ //not same track as before, check: same SuperParent or new one?
+    auto itTr=mSuperParents.find(partID) ;
+    if(itTr==mSuperParents.end()){
+      //Search parent
+      Int_t parentID = stack->GetCurrentTrack()->GetMother(0) ;
+      itTr=mSuperParents.find(parentID) ;
+      if(itTr==mSuperParents.end()){ //Neither track or its parent found: new SuperParent
+        mSuperParents[partID]=partID ;
+        superParent=partID ;
+        isNewPartile=true ;
+      }
+      else{ //parent found, this track - not
+        superParent=itTr->second ;
+        mSuperParents[partID]=superParent ;
+      }
+    }
+    else{
+      superParent=itTr->second ;      
+    }
+  }
+  else{
+    superParent=mCurentSuperParent ;
+  }
+
+
+  Double_t lostenergy = mcapp->Edep();
+  if (lostenergy < DBL_EPSILON && !isNewPartile)
+    return false; // do not create hits with zero energy deposition
+
+  if(!mGeom)
+    mGeom = Geometry::GetInstance();
+
+//  if(strcmp(mcapp->CurrentVolName(),"PXTL")!=0) //Non need to check, alwais there...
+//    return false ; //  We are not inside a PBWO crystal
+
+  Int_t  moduleNumber ;   
+  mcapp->CurrentVolOffID(10, moduleNumber) ; // get the PHOS module number ;
+  Int_t strip ;
+  mcapp->CurrentVolOffID(3, strip);
+  Int_t cell ;
+  mcapp->CurrentVolOffID(2, cell);
+  Int_t detID = mGeom->RelToAbsId(moduleNumber,strip,cell);
+
+  if(superParent==mCurentSuperParent && detID == mCurrentCellID && mCurrentHit) {
+    //continue with current hit
+    mCurrentHit->AddEnergyLoss(lostenergy);
+    return true ;
+  }
+
+  //try to find existing Hit
+  if(!isNewPartile){
+    for(Int_t itr=mHits->size()-1; itr>=0; itr--){
+      Hit *h = &(mHits->at(itr)) ;
+      if(h->GetTrackID()!=superParent) //switched to another SuperParent, do not search further
+        break ;
+      if(h->GetDetectorID() == detID){ //found correct hit
+        h->AddEnergyLoss(lostenergy);
+        mCurentSuperParent=superParent ;
+        mCurrentTrackID = partID ;
+        mCurrentCellID = detID ;
+        mCurrentHit = h;
+        return true ;
+      }
+    }
+
+  }
+  //Create new Hit
+  Double_t posX, posY, posZ, momX, momY, momZ, energy;
```

Comment:
I believe we can directly get posX etc as a float. This avoids type conversions.